src/Amadevus.RecordGenerator.Generators/Generation.cs

```diff
@@ -0,0 +1,56 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
+
+namespace Amadevus.RecordGenerator.Generators
+{
+    internal class Generation
+    {
+        public static readonly Generation Empty =
+            new Generation(TokenList(),
+                           ImmutableArray<BaseTypeSyntax>.Empty,
+                           ImmutableArray<MemberDeclarationSyntax>.Empty);
+
+        public SyntaxTokenList Modifiers { get; }
+        public ImmutableArray<BaseTypeSyntax> BaseTypes { get; }
+        public ImmutableArray<MemberDeclarationSyntax> Members;
+
+        Generation(SyntaxTokenList modifiers,
+                   ImmutableArray<BaseTypeSyntax> baseTypes,
+                   ImmutableArray<MemberDeclarationSyntax> members)
+        {
+            Members = members;
+            Modifiers = modifiers;
+            BaseTypes = baseTypes;
+        }
+
+        public Generation Add(Generation generation)
+            => generation is null ? throw new ArgumentNullException(nameof(generation))
+             : ReferenceEquals(generation, Empty) ? this
+             : ReferenceEquals(this, Empty) ? generation
+             : new Generation(Modifiers.AddRange(generation.Modifiers),
+                              BaseTypes.AddRange(generation.BaseTypes),
+                              Members.AddRange(generation.Members));
+
+        public Generation WithModifiers(SyntaxTokenList value) =>
+            new Generation(value, BaseTypes, Members);
+
+        public Generation WithBaseList(ImmutableArray<BaseTypeSyntax> value) =>
+            new Generation(Modifiers, value, Members);
+
+        public Generation WithMembers(ImmutableArray<MemberDeclarationSyntax> value) =>
+            new Generation(Modifiers, BaseTypes, value);
+
+        public Generation AddMember(MemberDeclarationSyntax member) =>
```

Comment:
The `params` version is not an overload since its named different (plural). This method was just to avoid an array allocation when you're adding just one member. Happy to go with just the `params` version. In fact, I'd be even inclined to add the following:

```
public GenerationResult AddMembers(MemberDeclarationSyntax member) =>
    AddMember(member);
```

This way, the array allocation is avoided without you having to think or worry. 😉

Reply:
Right. I'd say thought that this optimization is too early though. The allocations will happen anyway since we work on immutable nodes anyway, one allocation +/- won't impact performance (which is crazy slow anyway since we're doing a whole second compilation). However it's an additional similar method which doesn't *greatly* enhance comprehension/compression of the source code.

It can stay as is, I'm just academically discussing my POV.